File: src/php/core/src/Style.php
<?php

declare(strict_types=1);

namespace DiceBear;

use DiceBear\Error\StyleValidationError;
use DiceBear\Style\Canvas;
use DiceBear\Style\Color;
use DiceBear\Style\Component;
use DiceBear\Style\Meta;
use DiceBear\Validator\StyleValidator;

/**
 * Validated, lazily-decomposed wrapper around a style definition. Construction
 * runs the JSON Schema validator and stores a deep copy of the input so that
 * later mutation of the source object cannot leak into the rendered avatar.
 */
class Style
{
    /** @var array<string, mixed> */
    private array $data;
    private ?Meta $meta = null;
    private ?Canvas $canvas = null;
    /** @var array<string, Component>|null */
    private ?array $components = null;
    /** @var array<string, Color>|null */
    private ?array $colors = null;

    public function __construct(mixed $data)
    {
        StyleValidator::validate($data);

        $this->data = is_array($data) ? $data : json_decode(json_encode($data), true);

        $this->validateAliases();
    }

    /**
     * Returns the definition's `$id`, or `null` when not set.
     */
    public function id(): ?string
    {
        return $this->data['$id'] ?? null;
    }

    /**
     * Returns the definition's `$schema` URI, or `null` when not set.
     */
    public function schema(): ?string
    {
        return $this->data['$schema'] ?? null;
    }

    /**
     * Returns the definition's `$comment`, or `null` when not set.
     */
    public function comment(): ?string
    {
        return $this->data['$comment'] ?? null;
    }

    /**
     * Returns the {@see Meta} view, lazily constructed on first access.
     */
    public function meta(): Meta
    {
        return $this->meta ??= new Meta($this->data['meta'] ?? []);
    }

    /**
     * Returns the root SVG attributes from the definition, defaulting to an
     * empty array.
     *
     * @return array<string, mixed>
     */
    public function attributes(): array
    {
        return $this->data['attributes'] ?? [];
    }

    /**
     * Returns the {@see Canvas} view, lazily constructed on first access.
     */
    public function canvas(): Canvas
    {
        return $this->canvas ??= new Canvas($this->data['canvas']);
    }

    /**
     * Returns a name → {@see Component} map for all defined components, built
     * lazily on first access.
     *
     * @return array<string, Component>
     */
    public function components(): array
    {
        if ($this->components !== null) {
            return $this->components;
        }

        $entries = $this->data['components'] ?? [];
        $this->components = [];

        foreach ($entries as $name => $data) {
            if (!self::isAlias($data)) {
                $this->components[$name] = new Component($data);
            }
        }

        foreach ($entries as $name => $data) {
            if (self::isAlias($data)) {
                $this->components[$name] = new Component(
                    $data,
                    $this->components[$data['extends']] ?? null,
                );
            }
        }

        return $this->components;
    }

    /**
     * Verifies that every component declared via `extends` references an
     * existing, non-alias component in the same `components` map. The schema
     * itself cannot enforce cross-references between sibling keys.
     */
    private function validateAliases(): void
    {
        $components = $this->data['components'] ?? null;

        if ($components === null) {
            return;
        }

        $errors = [];

        foreach ($components as $name => $data) {
            if (!self::isAlias($data)) {
                continue;
            }

            $target = $data['extends'];
            $targetData = $components[$target] ?? null;

            if ($targetData === null) {
                $errors[] = [
                    'instancePath' => "/components/{$name}/extends",
                    'message' => "references unknown component \"{$target}\"",
                ];

                continue;
            }

            if (self::isAlias($targetData)) {
                $errors[] = [
                    'instancePath' => "/components/{$name}/extends",
                    'message' => "references alias \"{$target}\" — alias chains are not allowed",
                ];
            }
        }

        if (count($errors) > 0) {
            throw new StyleValidationError($errors);
        }
    }

    /**
     * @param array<string, mixed> $data
     */
    private static function isAlias(array $data): bool
    {
        return array_key_exists('extends', $data);
    }

    /**
     * Returns a name → {@see Color} map for all defined colors, built lazily
     * on first access.
     *
     * @return array<string, Color>
     */
    public function colors(): array
    {
        if ($this->colors === null) {
            $this->colors = [];

            foreach ($this->data['colors'] ?? [] as $name => $data) {
                $this->colors[$name] = new Color($data);
            }
        }

        return $this->colors;
    }
}

File: src/rust/core/src/style.rs
//! Validated, decomposed wrapper around a style definition.
//!
//! The per-concept views (canvas, element, component, color, meta) live in the
//! `style/` submodules, mirroring the original's `Style/` directory.

mod canvas;
mod color;
mod component;
mod component_translate;
mod component_variant;
mod element;
mod meta;
mod meta_creator;
mod meta_license;
mod meta_source;

pub(crate) use canvas::Canvas;
pub(crate) use color::Color;
pub(crate) use component::Component;
pub(crate) use element::{DynValue, Element};
pub(crate) use meta::Meta;

use indexmap::IndexMap;
use serde::Deserialize;
use serde_json::Value;

use crate::error::Error;
use crate::validate;
use component::RawComponent;

/// The raw, deserialized style definition.
#[derive(Deserialize)]
struct Definition {
    #[serde(rename = "$id")]
    id: Option<String>,
    #[serde(rename = "$schema")]
    schema: Option<String>,
    #[serde(rename = "$comment")]
    comment: Option<String>,
    meta: Option<Meta>,
    attributes: Option<IndexMap<String, DynValue>>,
    canvas: Canvas,
    components: Option<IndexMap<String, RawComponent>>,
    colors: Option<IndexMap<String, Color>>,
}

/// A parsed, validated style definition with its component aliases flattened.
pub struct Style {
    id: Option<String>,
    schema: Option<String>,
    comment: Option<String>,
    meta: Option<Meta>,
    attributes: Option<IndexMap<String, DynValue>>,
    canvas: Canvas,
    components: IndexMap<String, Component>,
    colors: IndexMap<String, Color>,
}

impl Style {
    /// Parses and validates a style definition from a JSON string.
    #[allow(clippy::should_implement_trait)]
    pub fn from_str(json: &str) -> Result<Self, Error> {
        Self::from_value(serde_json::from_str(json)?)
    }

    /// Validates and builds a style from an already-parsed JSON value.
    pub fn from_value(value: Value) -> Result<Self, Error> {
        validate::definition(&value)?;

        let def: Definition = serde_json::from_value(value)?;
        validate_aliases(def.components.as_ref())?;

        Ok(Self {
            id: def.id,
            schema: def.schema,
            comment: def.comment,
            meta: def.meta,
            attributes: def.attributes,
            canvas: def.canvas,
            components: component::build(def.components),
            colors: def.colors.unwrap_or_default(),
        })
    }

    /// Returns the definition's `$id`, or `None` when not set.
    pub fn id(&self) -> Option<&str> {
        self.id.as_deref()
    }

    /// Returns the definition's `$schema` URI, or `None` when not set.
    pub fn schema(&self) -> Option<&str> {
        self.schema.as_deref()
    }

    /// Returns the definition's `$comment`, or `None` when not set.
    pub fn comment(&self) -> Option<&str> {
        self.comment.as_deref()
    }

    pub(crate) fn canvas(&self) -> &Canvas {
        &self.canvas
    }

    pub(crate) fn root_attributes(&self) -> Option<&IndexMap<String, DynValue>> {
        self.attributes.as_ref()
    }

    pub(crate) fn meta(&self) -> Option<&Meta> {
        self.meta.as_ref()
    }

    pub(crate) fn components(&self) -> &IndexMap<String, Component> {
        &self.components
    }

    pub(crate) fn colors(&self) -> &IndexMap<String, Color> {
        &self.colors
    }
}

/// Verifies that every `extends` references an existing, non-alias component —
/// a cross-key constraint the JSON Schema cannot express.
fn validate_aliases(components: Option<&IndexMap<String, RawComponent>>) -> Result<(), Error> {
    let Some(components) = components else {
        return Ok(());
    };

    let mut errors: Vec<String> = Vec::new();

    for (name, entry) in components {
        let RawComponent::Alias { extends } = entry else {
            continue;
        };

        match components.get(extends) {
            None => errors.push(format!(
                "/components/{name}/extends references unknown component \"{extends}\""
            )),
            Some(RawComponent::Alias { .. }) => errors.push(format!(
                "/components/{name}/extends references alias \"{extends}\" — alias chains are not allowed"
            )),
            Some(RawComponent::Base(_)) => {}
        }
    }

    if errors.is_empty() {
        Ok(())
    } else {
        Err(Error::Validation(errors.join("; ")))
    }
}
